src/geom/domain/boundary.rs

```diff
@@ -0,0 +1,584 @@
+use crate::{
+    access, clone, fmt_report, geom::{plane::ray_plane_intersection, Cube, Hit, Ray, Side, Trace}, math::{Dir3, Point3, Vec3}, phys::{Photon, Reflectance}, sim::Attribute
+};
+use rand::rngs::ThreadRng;
+use std::fmt::{Display, Formatter};
+
+/// Struct that represents a boundary.
+/// This will be used to determine how the boundary conditions behaves when it interacts
+/// with photon packets.
+pub struct Boundary {
+    bounding_box: Cube,
+    top: BoundaryCondition,
+    bottom: BoundaryCondition,
+    north: BoundaryCondition,
+    east: BoundaryCondition,
+    south: BoundaryCondition,
+    west: BoundaryCondition,
+}
+
+impl Boundary {
+    access!(bounding_box, bounding_box_mut: Cube);
+    access!(top, top_mut: BoundaryCondition);
+    access!(bottom, bottom_mut: BoundaryCondition);
+    access!(north, north_mut: BoundaryCondition);
+    access!(east, east_mut: BoundaryCondition);
+    access!(south, south_mut: BoundaryCondition);
+    access!(west, west_mut: BoundaryCondition);
+
+    pub fn new(bounding_box: Cube) -> Self {
+        Self {
+            bounding_box,
+            top: BoundaryCondition::default(),
+            bottom: BoundaryCondition::default(),
+            north: BoundaryCondition::default(),
+            east: BoundaryCondition::default(),
+            south: BoundaryCondition::default(),
+            west: BoundaryCondition::default(),
+        }
+    }
+
+    pub fn new_kill(bounding_box: Cube) -> Self {
+        Self {
+            bounding_box,
+            top: BoundaryCondition::Kill,
+            bottom: BoundaryCondition::Kill,
+            north: BoundaryCondition::Kill,
+            east: BoundaryCondition::Kill,
+            south: BoundaryCondition::Kill,
+            west: BoundaryCondition::Kill,
+        }
+    }
+
+    pub fn new_reflect(bounding_box: Cube, reflect: Reflectance) -> Self {
+        Self {
+            bounding_box,
+            top: BoundaryCondition::Reflect(reflect.clone()),
+            bottom: BoundaryCondition::Reflect(reflect.clone()),
+            north: BoundaryCondition::Reflect(reflect.clone()),
+            east: BoundaryCondition::Reflect(reflect.clone()),
+            south: BoundaryCondition::Reflect(reflect.clone()),
+            west: BoundaryCondition::Reflect(reflect.clone()),
+        }
+    }
+
+    pub fn new_periodic(bounding_box: Cube, padding: f64) -> Self {
+        Self {
+            bounding_box,
+            top: BoundaryCondition::Periodic(padding),
+            bottom: BoundaryCondition::Periodic(padding),
+            north: BoundaryCondition::Periodic(padding),
+            east: BoundaryCondition::Periodic(padding),
+            south: BoundaryCondition::Periodic(padding),
+            west: BoundaryCondition::Periodic(padding),
+        }
+    }
```

Comment:
Don't necessarily mind this - i quite like how explicit the interface is for creating boundaries, It feels like code itself can probably be trimmed down but at the moment definitely keep it how it is :)

Reply:
Agreed! It's quite verbose, but it is intended as a first working version that provides the flexibility for the MPI implementation to follow. More than happy to revisit this in quality-of-life improvements in the future.